cli/napi/uv.rs

```diff
@@ -0,0 +1,237 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+
+use deno_runtime::deno_napi::*;
+use std::mem::MaybeUninit;
+use std::ptr::addr_of_mut;
+
+#[allow(clippy::print_stderr)]
+fn assert_ok(res: c_int) -> c_int {
+  if res != 0 {
+    eprintln!("bad result in uv polyfill: {res}");
+    // don't panic because that might unwind into
+    // c/c++
+    std::process::abort();
+  }
+  res
+}
+
+use crate::napi::js_native_api::napi_create_string_utf8;
+use crate::napi::node_api::napi_create_async_work;
+use crate::napi::node_api::napi_delete_async_work;
+use crate::napi::node_api::napi_queue_async_work;
+use std::ffi::c_int;
+
+#[cfg(unix)]
+mod mutex {
+  use super::*;
+  type uv_mutex_t = libc::pthread_mutex_t;
+  #[no_mangle]
+  unsafe extern "C" fn uv_mutex_init(lock: *mut uv_mutex_t) -> c_int {
+    use std::mem::MaybeUninit;
+    let mut attr = MaybeUninit::<libc::pthread_mutexattr_t>::uninit();
+    unsafe {
+      assert_ok(libc::pthread_mutexattr_init(attr.as_mut_ptr()));
+      let mut attr = attr.assume_init();
+      let attr_ptr = addr_of_mut!(attr);
+      assert_ok(libc::pthread_mutexattr_settype(
+        attr_ptr,
+        libc::PTHREAD_MUTEX_ERRORCHECK,
+      ));
+      let err = libc::pthread_mutex_init(lock, attr_ptr);
+      assert_ok(libc::pthread_mutexattr_destroy(attr_ptr));
+      if libc::EDOM > 0 {
+        err
+      } else {
+        -err
+      }
+    }
+  }
+
+  #[no_mangle]
+  unsafe extern "C" fn uv_mutex_lock(lock: *mut uv_mutex_t) {
```

Comment:
these could be written in terms of `std::sync::Mutex` and gain safety and platform compat.

Reply:
Done, though I went with `parking_lot::Mutex` instead to avoid having to store the guard with a fake lifetime.